db/compaction/compaction.cc

```diff
@@ -301,6 +302,68 @@ Compaction::Compaction(
     }
   }
   output_split_key_ = temp_split_key;
+
+  PopulatePrecludeLastOutputBoundaries();
+}
+
+void Compaction::PopulatePrecludeLastOutputBoundaries() {
+  if (!SupportsPrecludeLast()) {
+    return;
+  }
+
+  const Comparator* ucmp =
+      input_vstorage_->InternalComparator()->user_comparator();
+
+  std::vector<CompactionBoundary> all_boundaries;
+
+  assert(preclude_last_output_level_ > 0);
+
+  // find the all the valid preclude_last key ranges, which is the input
+  // files' ranges, merge them if they're overlapped.
+  for (const auto& input : inputs_) {
+    if (input.level > preclude_last_output_level_ || input.files.empty()) {
+      continue;
+    }
+    for (const auto& file : input.files) {
+      all_boundaries.emplace_back(&file->smallest, &file->largest);
+    }
+  }
```

Comment:
We should only write to the ranges that are locked in the `preclude_last_output_level_`, either because a file exists for that range in `preclude_last_output_level_` or because a compaction has been registered for that range. I don't think the range of a file on a higher input level would be prevented from being written to `preclude_last_output_level_` by another compaction/ingestion.

Reply:
By "registered" I mean https://github.com/facebook/rocksdb/blob/e6c5e0ab9a7235a90a511553cb3c5fb2cec85e99/db/compaction/compaction_picker.cc#L285-L299 is aware of ranges on `preclude_last_output_level_` that may be written.

Reply:
the same comment as above. Let me know if it's a different issue.